File: Source/HiGame/Private/HiAbilities/HiCollisionLibrary.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "HiAbilities/HiCollisionLibrary.h"

#include "AbilitySystemLog.h"
#include "DrawDebugHelpers.h"
#include "Kismet/KismetSystemLibrary.h"
#include "HiAbilities/HiMountActor.h"
#include "PhysicsEngine/BodyInstance.h"
#include "PhysicalMaterials/PhysicalMaterial.h"
#include "HiAbilities/HiTraceUtils.h"

UHiCollisionLibrary::UHiCollisionLibrary()
{
}

UHiCollisionLibrary::~UHiCollisionLibrary()
{
}

void UHiCollisionLibrary::PerformOverlapActorsBP(UObject* WorldContextObject, const FHiTargetActorSpec& TargetActorSpec, const FVector& Origin, const FVector& ForwardVector, FGameplayTargetDataFilterHandle Filter, TArray<AActor*>& HitActors, bool& NeedDestroy, bool bDebug)
{
	TArray<TWeakObjectPtr<AActor>> WOutActors;
	UHiCollisionLibrary::PerformOverlapActors(WorldContextObject, TargetActorSpec, Origin, ForwardVector, Filter, WOutActors, NeedDestroy, bDebug);
	
	for (auto OutActor : WOutActors)
	{
		AActor* Actor = OutActor.Get();
		HitActors.Add(Actor);
	}
}

void UHiCollisionLibrary::PerformOverlapActors(UObject* WorldContextObject, const FHiTargetActorSpec& TargetActorSpec, const FVector& Origin, const FVector& ForwardVector, FGameplayTargetDataFilterHandle Filter, TArray<TWeakObjectPtr<AActor>>& HitActors, bool& NeedDestroy, bool bDebug, float LifeTime)
{
	TArray<AActor*> OutActors;
	TArray<AActor*> ActorsToIgnore;
	TArray<TEnumAsByte<EObjectTypeQuery>> ObjectTypes;
	// TODO default to Pawn and MountActor.
	ObjectTypes.Add(EObjectTypeQuery::ObjectTypeQuery3);
	ObjectTypes.Add(EObjectTypeQuery::ObjectTypeQuery7);
	const float AngleRad = FMath::DegreesToRadians(TargetActorSpec.Angle);

	switch(TargetActorSpec.CalcRangeType)
	{
	case Circle:
		SphereOverlapActors(WorldContextObject, ObjectTypes, Origin, TargetActorSpec.Radius, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nullptr, ActorsToIgnore, OutActors, bDebug, LifeTime);
		break;
		
	case Section:
		SectionOverlapActors(WorldContextObject, ObjectTypes, Origin, ForwardVector, TargetActorSpec.Radius, AngleRad, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nullptr, ActorsToIgnore, OutActors, bDebug, LifeTime);
		break;

	case CalcRangeTypeRect:
		BoxOverlapActors(WorldContextObject, ObjectTypes, Origin, ForwardVector, TargetActorSpec.Length, TargetActorSpec.HalfWidth, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nullptr, ActorsToIgnore, OutActors, false, bDebug, LifeTime);
		break;

	default:
		break;
	}

	for (int32 i = 0; i < OutActors.Num(); ++i)
	{
		// TODO consider hit crushable box.
		AActor* Actor = OutActors[i];
		APawn* PawnActor = Cast<APawn>(OutActors[i]);
		AHiMountActor* HiMountActor = Cast<AHiMountActor>(OutActors[i]);
		if (!PawnActor && !HiMountActor)
		{
			// Collision with scene.
			NeedDestroy = true;
			continue;
		}
		
		if (!HitActors.Contains(Actor) && Filter.FilterPassesForActor(Actor))
		{
			// Limit one calculation target limit.
			// TODO priority.
			if (HitActors.Num() >= TargetActorSpec.CalcTargetLimit)
			{
				return;
			}
			
			HitActors.Add(Actor);
		}
	}
}

bool UHiCollisionLibrary::SphereOverlapActors(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, float Radius, float UpHeight, float DownHeight, UClass* ActorClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<AActor*>& OutActors, bool bDebug, float LifeTime)
{
	OutActors.Empty();

	TArray<UPrimitiveComponent*> OverlapComponents;
	bool bOverlapped = SphereOverlapComponents(WorldContextObject, ObjectTypes, Origin, Radius, UpHeight, DownHeight, NULL, ActorsToIgnore, OverlapComponents, bDebug, LifeTime);
	if (bOverlapped)
	{
		UKismetSystemLibrary::GetActorListFromComponentList(OverlapComponents, ActorClassFilter, OutActors);
	}

	return (OutActors.Num() > 0);
}


bool UHiCollisionLibrary::SphereOverlapComponents(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, float Radius, float UpHeight, float DownHeight, UClass* ComponentClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<UPrimitiveComponent*>& OutComponents, bool bDebug, float LifeTime)
{
	UKismetSystemLibrary::SphereOverlapComponents(WorldContextObject, Origin, Radius, ObjectTypes, ComponentClassFilter, ActorsToIgnore, OutComponents);

	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::LogAndReturnNull);
	if (World != nullptr)
	{
#if ENABLE_DRAW_DEBUG
		if (bDebug)
		{
			DrawDebugSphere(World, Origin, Radius, 16, FColor::Red, false, LifeTime);
		}
#endif // ENABLE_DRAW_DEBUG
	}

	// Check in up and down height.
	for (auto It = OutComponents.CreateIterator(); It; ++ It)
	{
		auto const& CurComponent = *It;
		if ((UpHeight > 0 || DownHeight > 0) && !CheckInHeightZ(Origin.Z, CurComponent->Bounds.Origin.Z + CurComponent->Bounds.BoxExtent.Z, CurComponent->Bounds.Origin.Z - CurComponent->Bounds.BoxExtent.Z, UpHeight, DownHeight))
		{
			It.RemoveCurrent();
		}
	}

	return (OutComponents.Num() > 0);
}

bool UHiCollisionLibrary::SectionOverlapActors(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, const FVector& ForwardVector, float Radius, float Angle, float UpHeight, float DownHeight, UClass* ActorClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<AActor*>& OutActors, bool bDebug, float LifeTime)
{
	OutActors.Empty();

	TArray<UPrimitiveComponent*> OverlapComponents;
	bool bOverlapped = SectionOverlapComponents(WorldContextObject, ObjectTypes, Origin, ForwardVector, Radius, Angle, UpHeight, DownHeight, NULL, ActorsToIgnore, OverlapComponents, bDebug, LifeTime);
	if (bOverlapped)
	{
		UKismetSystemLibrary::GetActorListFromComponentList(OverlapComponents, ActorClassFilter, OutActors);
	}

	return (OutActors.Num() > 0);
}

bool UHiCollisionLibrary::SectionOverlapComponents(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, const FVector& ForwardVector, float Radius, float Angle, float UpHeight, float DownHeight, UClass* ComponentClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<UPrimitiveComponent*>& OutComponents, bool bDebug, float LifeTime)
{
	SphereOverlapComponents(WorldContextObject, ObjectTypes, Origin, Radius, UpHeight, DownHeight, ComponentClassFilter, ActorsToIgnore, OutComponents);

	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::LogAndReturnNull);
	if (World != nullptr)
	{
#if ENABLE_DRAW_DEBUG
		if (bDebug)
		{
			DrawDebugPoint(World, Origin, 10, FColor::Green, false, LifeTime);
			DrawDebugLine(World, Origin, Origin + ForwardVector * Radius, FColor::Yellow, false, LifeTime);
			DrawDebugCircleArc(World, Origin, Radius, ForwardVector, Angle/2, 8, FColor::Green, false, LifeTime);
			DrawDebugSphere(World, Origin, Radius, 16, FColor::Red, false, LifeTime);
		}
#endif // ENABLE_DRAW_DEBUG
	}

	// After sphere overlap, check angle in section.
	for (auto It = OutComponents.CreateIterator(); It; ++ It)
	{
		const UPrimitiveComponent* CurComp = (*It);
		if (!CheckInSection(CurComp->GetComponentLocation(), Origin, ForwardVector, Angle))
		{
			const auto CurComponent = *It;
			It.RemoveCurrent();
		}
	}
	
	return (OutComponents.Num() > 0);
}

bool UHiCollisionLibrary::BoxOverlapActors(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, const FVector& ForwardVector, float Length, float HalfWidth, float UpHeight, float DownHeight, UClass* ActorClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<AActor*>& OutActors, bool bKeepOrigin, bool bDebug, float LifeTime)
{
	OutActors.Empty();

	TArray<UPrimitiveComponent*> OverlapComponents;
	bool bOverlapped = BoxOverlapComponents(WorldContextObject, ObjectTypes, Origin, ForwardVector, Length, HalfWidth, UpHeight, DownHeight, NULL, ActorsToIgnore, OverlapComponents, bKeepOrigin, bDebug, LifeTime);
	if (bOverlapped)
	{
		UKismetSystemLibrary::GetActorListFromComponentList(OverlapComponents, ActorClassFilter, OutActors);
	}

	return (OutActors.Num() > 0);
}

bool UHiCollisionLibrary::BoxOverlapComponents(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector& Origin, const FVector& ForwardVector, float Length, float HalfWidth, float UpHeight, float DownHeight, UClass* ComponentClassFilter, const TArray<AActor*>& ActorsToIgnore, TArray<UPrimitiveComponent*>& OutComponents, bool bKeepOrigin, bool bDebug, float LifeTime)
{
	OutComponents.Empty();

	const FQuat& Rot = ForwardVector.Rotation().Quaternion();
	const FVector BoxExtent = FVector( Length / 2, HalfWidth, FMath::Max(UpHeight, DownHeight));
	FVector NewOrigin = Origin;
	if (!bKeepOrigin)
	{
		NewOrigin = Origin + ForwardVector * Length / 2;
	}

	FCollisionQueryParams Params(SCENE_QUERY_STAT(BoxOverlapComponents), false);
	Params.AddIgnoredActors(ActorsToIgnore);

	TArray<FOverlapResult> Overlaps;
	FCollisionObjectQueryParams ObjectParams;
	for (auto Iter = ObjectTypes.CreateConstIterator(); Iter; ++Iter)
	{
		const ECollisionChannel & Channel = UCollisionProfile::Get()->ConvertToCollisionChannel(false, *Iter);
		ObjectParams.AddObjectTypesToQuery(Channel);
	}

	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::LogAndReturnNull);
	if (World != nullptr)
	{
		World->OverlapMultiByObjectType(Overlaps, NewOrigin, Rot, ObjectParams, FCollisionShape::MakeBox(BoxExtent), Params);

#if ENABLE_DRAW_DEBUG
		if (bDebug)
		{
			DrawDebugPoint(World, NewOrigin, 10, FColor::Green, false, LifeTime);
			DrawDebugLine(World, NewOrigin, NewOrigin + ForwardVector * Length, FColor::Yellow, false, LifeTime);
			if (!bKeepOrigin)
			{
				DrawDebugPoint(World, NewOrigin, 10, FColor::Red, false, LifeTime);
			}
			DrawDebugBox(World, NewOrigin, BoxExtent, Rot, FColor::Red, false, LifeTime);
		}
#endif // ENABLE_DRAW_DEBUG
	}

	for (int32 OverlapIdx=0; OverlapIdx<Overlaps.Num(); ++OverlapIdx)
	{
		FOverlapResult const& O = Overlaps[OverlapIdx];
		auto OverlapComp = O.Component;

		if (! OverlapComp.IsValid())
		{
			continue;
		}
		
		if (! CheckInHeightZ(NewOrigin.Z, OverlapComp->Bounds.Origin.Z + OverlapComp->Bounds.BoxExtent.Z, OverlapComp->Bounds.Origin.Z - OverlapComp->Bounds.BoxExtent.Z, UpHeight, DownHeight))
		{
			continue;
		}
		
		if ( !ComponentClassFilter || OverlapComp.Get()->IsA(ComponentClassFilter) )
		{
			OutComponents.Add(OverlapComp.Get());
		}
	}

	return (OutComponents.Num() > 0);
}

bool UHiCollisionLibrary::CheckInHeight(const FVector& TargetPos, const FVector& Origin, float UpHeight, float DownHeight)
{
	float HeightDelta = TargetPos.Z - Origin.Z;
	if ((DownHeight > 0 && HeightDelta < -DownHeight) || (UpHeight > 0 && HeightDelta > UpHeight))
	{
		return false;
	}
	
	return true;
}

bool UHiCollisionLibrary::CheckInHeightZ(float OriginZ, float MaxZ, float MinZ, float UpHeight, float DownHeight)
{
	if ((UpHeight > 0 && MaxZ < OriginZ - DownHeight) || (DownHeight > 0 && MinZ > OriginZ + UpHeight))
	{
		return false;
	}

	return true;
}

bool UHiCollisionLibrary::CheckInSection(const FVector& TargetPos, const FVector& Origin, const FVector& ForwardVector, float Radians)
{
	FVector TargetPosWithoutZ = FVector(TargetPos.X, TargetPos.Y, 0);
	FVector OriginPosWithoutZ = FVector(Origin.X, Origin.Y, 0);
	const FVector OriginToTarget = TargetPosWithoutZ - OriginPosWithoutZ;
	const float CosHalfAngle = FMath::Cos(FMath::Clamp(Radians/2, 0.f, PI));
	if(FVector::DotProduct(OriginToTarget.GetUnsafeNormal(), ForwardVector) > CosHalfAngle)
	{
		return true;
	}

	return false;
}

bool UHiCollisionLibrary::CheckInDirectionBySection(const FVector& TargetPos, const FVector& Origin, const FVector& ForwardVector, int StartAngle, int EndAngle)
{
	StartAngle = FMath::Clamp(StartAngle, 0, 360);
	EndAngle = FMath::Clamp(EndAngle, 0, 360);
	StartAngle = FMath::Min(StartAngle, EndAngle);

	if (StartAngle < 180 && EndAngle > 180)
	{
		// 跨越180度得分开判断
		return CheckInDirectionBySection(TargetPos, Origin, ForwardVector, StartAngle, 180) || CheckInDirectionBySection(TargetPos, Origin, ForwardVector, 180, EndAngle);
	}
	
	const FVector TargetPosWithoutZ = FVector(TargetPos.X, TargetPos.Y, 0);
	const FVector OriginPosWithoutZ = FVector(Origin.X, Origin.Y, 0);
	const FVector OriginToTarget = TargetPosWithoutZ - OriginPosWithoutZ;

	auto const DotForward = FVector::DotProduct(OriginToTarget.GetUnsafeNormal(), ForwardVector.GetUnsafeNormal());
	if (DotForward < 0.0f && EndAngle <= 180)
	{
		// DotForward < 0说明Target在Origin的后方
		return false;
	}
	else if (DotForward > 0.0f && EndAngle > 180)
	{
		// DotForward > 0 说明Target在Origin的前方
		return false;
	}
	
	const FVector ZNormal(0, 0, 1);
	const FVector RightForwardVector = FVector::CrossProduct(ZNormal, ForwardVector);
	auto const DotRightForward = FVector::DotProduct(OriginToTarget.GetUnsafeNormal(), RightForwardVector.GetUnsafeNormal());

	auto const StartRadians = FMath::DegreesToRadians(StartAngle);
	auto const EndRadians = FMath::DegreesToRadians(EndAngle);
	
	auto const StartCos = FMath::Cos(StartRadians);
	auto const EndCos = FMath::Cos(EndRadians);
	
	bool ret = false;
	if (EndAngle <= 180)
	{
		// EndAngle <= 180时，Angle越大，cos越小
		ret = (DotRightForward < StartCos || FMath::Abs(DotRightForward - StartCos) < 0.00001f) && // DotRightForward <= StartCos
			(DotRightForward > EndCos || FMath::Abs(DotRightForward - EndCos) < 0.00001f); // DotRightForward >= EndCos
	}
	else
	{
		// EndAngle > 180时，Angle越大，cos越大
		ret = (DotRightForward > StartCos || FMath::Abs(DotRightForward - StartCos) < 0.00001f) && // DotRightForward >= StartCos
			(DotRightForward < EndCos || FMath::Abs(DotRightForward - EndCos) < 0.00001f); // DotRightForward <= EndCos
	}
	
	return ret;
}

bool UHiCollisionLibrary::ComponentTraceMulti(const UObject* WorldContextObject, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, const FVector Start, const FVector End, const FRotator Rotation, UPrimitiveComponent* PrimComp, TArray<FHitResult>& OutHits)
{
	if (!WorldContextObject || !PrimComp)
	{
		return false;
	}
	
	AActor* const Actor = PrimComp->GetOwner();
	static const FName TraceTagName = TEXT("HiCollisionLibrary");
	FComponentQueryParams Params(SCENE_QUERY_STAT(MoveComponent), Actor);
	FCollisionResponseParams ResponseParam;
	PrimComp->InitSweepCollisionParams(Params, ResponseParam);
	Params.bIgnoreTouches |= !(PrimComp->GetGenerateOverlapEvents());
	Params.TraceTag = TraceTagName;
	UWorld* MyWorld = WorldContextObject->GetWorld();

	FCollisionObjectQueryParams ObjectParams;
	for (auto Iter = ObjectTypes.CreateConstIterator(); Iter; ++Iter)
	{
		const ECollisionChannel & Channel = UCollisionProfile::Get()->ConvertToCollisionChannel(false, *Iter);
		ObjectParams.AddObjectTypesToQuery(Channel);
	}
	
	return MyWorld ? MyWorld->ComponentSweepMulti(OutHits, PrimComp, Start, End, Rotation, Params, ObjectParams) : false;
}

bool UHiCollisionLibrary::CapsuleTraceMultiForObjects(const UObject* WorldContextObject, const FVector Start, const FVector End, const FRotator Orientation, float Radius, float HalfHeight, const TArray<TEnumAsByte<EObjectTypeQuery>>& ObjectTypes, bool bTraceComplex, const TArray<AActor*>& ActorsToIgnore, EDrawDebugTrace::Type DrawDebugType, TArray<FHitResult>& OutHits, bool bIgnoreSelf, FLinearColor TraceColor, FLinearColor TraceHitColor, float DrawTime)
{
	static const FName CapsuleTraceMultiName(TEXT("CapsuleTraceMultiForObjects"));
	FCollisionQueryParams Params = HiConfigureCollisionParams(CapsuleTraceMultiName, bTraceComplex, ActorsToIgnore, bIgnoreSelf, WorldContextObject);

	FCollisionObjectQueryParams ObjectParams = HiConfigureCollisionObjectParams(ObjectTypes);
	if (ObjectParams.IsValid() == false)
	{
		UE_LOG(LogBlueprintUserMessages, Warning, TEXT("Invalid object types"));
		return false;
	}

	UWorld* World = GEngine->GetWorldFromContextObject(WorldContextObject, EGetWorldErrorMode::LogAndReturnNull);
	bool const bHit = World ? World->SweepMultiByObjectType(OutHits, Start, End, Orientation.Quaternion(), ObjectParams, FCollisionShape::MakeCapsule(Radius, HalfHeight), Params) : false;

#if ENABLE_DRAW_DEBUG
	HiDrawDebugCapsuleTraceMulti(World, Start, End, Orientation, Radius, HalfHeight, DrawDebugType, bHit, OutHits, TraceColor, TraceHitColor, DrawTime);
#endif

	return bHit;
}

void UHiCollisionLibrary::GetComponentPhysicsMaterial(UPrimitiveComponent* PrimComp, bool bSimpleCollision, TArray<UPhysicalMaterial*> &OutPhysMaterials)
{	
	if(PrimComp)
	{
		if (bSimpleCollision)
		{
			UPhysicalMaterial* PhysMat = PrimComp->BodyInstance.GetSimplePhysicalMaterial();
			if (PhysMat)
			{
				OutPhysMaterials.Add(PhysMat);	
			}			
		}
		else
		{
			PrimComp->BodyInstance.GetComplexPhysicalMaterials(OutPhysMaterials);
		}
	}	
}

File: HiCode/HiScript/common/HiCollisionLibrary.lua
require "UnLua"

local G = require("G")
local switches = require("switches")

local SkillUtils = require("common.skill_utils")

HiCollisionLibrary = {}

function HiCollisionLibrary:InitCollisionObjectTypes()
     self.DefaultCollisionObjectTypes = UE.TArray(UE.EObjectTypeQuery)
     self.DefaultCollisionObjectTypes:Add(UE.EObjectTypeQuery.Pawn)
     self.DefaultCollisionObjectTypes:Add(UE.EObjectTypeQuery.MountActor)
     self.DefaultCollisionObjectTypes:Add(UE.EObjectTypeQuery.Destructible)
     self.DefaultCollisionObjectTypes:Add(UE.EObjectTypeQuery.PhysicsBody)
     -- self.DefaultCollisionObjectTypes:Add(UE.EObjectTypeQuery.Blast)
end

HiCollisionLibrary:InitCollisionObjectTypes()

-- Perform overlaps with specified target actor spec.
function HiCollisionLibrary.PerformOverlapComponents(WorldContextObject, TargetActorSpec, Origin, ForwardVector, ObjectTypes, ActorsToIgnore, Hits, bDebug)
    if not switches.TraceDebug then
        bDebug = false
    end

    local CalcRangeType = TargetActorSpec.CalcRangeType
    local OutComponents = UE.TArray(UE.UPrimitiveComponent)

    if not ObjectTypes or ObjectTypes:Length() == 0 then
        ObjectTypes = HiCollisionLibrary.DefaultCollisionObjectTypes
    end

    if CalcRangeType == Enum.Enum_CalcRangeType.Circle then
        UE.UHiCollisionLibrary.SphereOverlapComponents(WorldContextObject, ObjectTypes, Origin, TargetActorSpec.Radius, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nil, ActorsToIgnore, OutComponents, bDebug)
    elseif CalcRangeType == Enum.Enum_CalcRangeType.Section then
        local SectionRadian = UE.UKismetMathLibrary.DegreesToRadians(TargetActorSpec.Angle)
        UE.UHiCollisionLibrary.SectionOverlapComponents(WorldContextObject, ObjectTypes, Origin, ForwardVector, TargetActorSpec.Radius, SectionRadian, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nil, ActorsToIgnore, OutComponents, bDebug)
    elseif CalcRangeType == Enum.Enum_CalcRangeType.Rect then
        UE.UHiCollisionLibrary.BoxOverlapComponents(WorldContextObject, ObjectTypes, Origin, ForwardVector, TargetActorSpec.Length, TargetActorSpec.HalfWidth, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight, nil, ActorsToIgnore, OutComponents, false, bDebug)
    else
        return
    end

    -- Filter and limit single calc count.
    local OutHits = SkillUtils.MakeHitResultsFromComponents(OutComponents, Origin)
    Hits:Append(OutHits)
end

function HiCollisionLibrary.PerformOverlapActors(WorldContextObject, Origin, ForwardVector, RangeType, Radius, Angle, Length, HalfWidth, UpHeight, DownHeight, ObjectTypes, bDebug, LifeTime)
    if not switches.TraceDebug then
        bDebug = false
    end

    local ActorsToIgnore = UE.TArray(UE.AActor)
    local OutActors = UE.TArray(UE.AActor)

    if not ObjectTypes or ObjectTypes:Length() == 0 then
        ObjectTypes = HiCollisionLibrary.DefaultCollisionObjectTypes
    end

    if RangeType == Enum.Enum_CalcRangeType.Circle then
        UE.UHiCollisionLibrary.SphereOverlapActors(WorldContextObject, ObjectTypes, Origin, Radius, UpHeight, DownHeight, nil, ActorsToIgnore, OutActors, bDebug, LifeTime)
    elseif RangeType == Enum.Enum_CalcRangeType.Section then
        local SectionRadian = UE.UKismetMathLibrary.DegreesToRadians(Angle)
        UE.UHiCollisionLibrary.SectionOverlapActors(WorldContextObject, ObjectTypes, Origin, ForwardVector, Radius, SectionRadian, UpHeight, DownHeight, nil, ActorsToIgnore, OutActors, bDebug, LifeTime)
    elseif RangeType == Enum.Enum_CalcRangeType.Rect then
        UE.UHiCollisionLibrary.BoxOverlapActors(WorldContextObject, ObjectTypes, Origin, ForwardVector, Length, HalfWidth, UpHeight, DownHeight, nil, ActorsToIgnore, OutActors, false, bDebug, LifeTime)
    end

    return OutActors
end

function HiCollisionLibrary.PerformSweep(WorldContextObject, TargetActorSpec, OriginLocation, OriginRotation, SweepDis, ForwardVector, ObjectTypes, ActorsToIgnore, Hits, DebugType)
    if not switches.TraceDebug then
        DebugType = UE.EDrawDebugTrace.None
    end
    
    local CalcRangeType = TargetActorSpec.CalcRangeType
    if not ActorsToIgnore then
        ActorsToIgnore = UE.TArray(UE.AActor)
    end
    local OutHits = UE.TArray(UE.FHitResult)
    if not SweepDis or SweepDis <= 0 then
        SweepDis = 20
    end
    local EndLocation = OriginLocation + ForwardVector * SweepDis

    if not ObjectTypes or ObjectTypes:Length() == 0 then
        ObjectTypes = HiCollisionLibrary.DefaultCollisionObjectTypes
    end

    if CalcRangeType == Enum.Enum_CalcRangeType.Circle then
        UE.UKismetSystemLibrary.SphereTraceMultiForObjects(WorldContextObject, OriginLocation, EndLocation, TargetActorSpec.Radius, ObjectTypes, true, ActorsToIgnore, DebugType, OutHits, true)
    elseif CalcRangeType == Enum.Enum_CalcRangeType.Rect then
        local HalfSize = UE.FVector(TargetActorSpec.Length / 2, TargetActorSpec.HalfWidth, (TargetActorSpec.UpHeight + TargetActorSpec.DownHeight) / 2);
        local BoxForwardVector = UE.UKismetMathLibrary.GetForwardVector(OriginRotation)
        BoxForwardVector.Z = 0
        BoxForwardVector = UE.UKismetMathLibrary.Normal(BoxForwardVector)
        local BoxLocation = UE.FVector(OriginLocation.X, OriginLocation.Y, TargetActorSpec.UpHeight + OriginLocation.Z - (TargetActorSpec.UpHeight + TargetActorSpec.DownHeight) / 2) + BoxForwardVector * TargetActorSpec.Length / 2
        EndLocation = BoxLocation + ForwardVector * SweepDis
        if DebugType == UE.EDrawDebugTrace.ForDuration then
            UE.UKismetSystemLibrary.DrawDebugPoint(WorldContextObject, OriginLocation, 30, UE.FLinearColor(1, 0, 0), 30)
            UE.UKismetSystemLibrary.DrawDebugArrow(WorldContextObject, BoxLocation, EndLocation, 5, UE.FLinearColor(0, 1, 0), 30, 5)
        end
        UE.UKismetSystemLibrary.BoxTraceMultiForObjects(WorldContextObject, BoxLocation, EndLocation, HalfSize, OriginRotation, ObjectTypes, true, ActorsToIgnore, DebugType, OutHits, true)
    elseif CalcRangeType == Enum.Enum_CalcRangeType.Section then
        UE.UKismetSystemLibrary.SphereTraceMultiForObjects(WorldContextObject, OriginLocation, EndLocation, TargetActorSpec.Radius, ObjectTypes, true, ActorsToIgnore, DebugType, OutHits, true)

        -- Check in angle.
        local Ind = 1
        while Ind <= OutHits:Length() do
            local HitResult = OutHits:Get(Ind)
            if not UE.UHiCollisionLibrary.CheckInSection(HitResult.ImpactPoint, OriginLocation, ForwardVector, UE.UKismetMathLibrary.DegreesToRadians(TargetActorSpec.Angle)) then
                OutHits:Remove(Ind)
            else
                Ind = Ind + 1
            end
        end
    end

    -- Handle TraceMulti return same components multi times.
    OutHits = UniqueHits(OutHits)

    -- Check in height.
    local Ind = 1
    while Ind <= OutHits:Length() do
        local HitResult = OutHits:Get(Ind)
        local CurComp = HitResult.Component
        local CompLoc, CompBounds = UE.UKismetSystemLibrary.GetComponentBounds(CurComp)
        if not UE.UHiCollisionLibrary.CheckInHeightZ(OriginLocation.Z, CompLoc.Z + CompBounds.Z, CompLoc.Z - CompBounds.Z, TargetActorSpec.UpHeight, TargetActorSpec.DownHeight) then
            OutHits:Remove(Ind)
        else
            Ind = Ind + 1
        end
    end

    Hits:Append(OutHits)
end

function UniqueHits(Hits)
    local OutHits = UE.TArray(UE.FHitResult)
    local Components = UE.TSet(UE.UPrimitiveComponent)
    for Ind = 1, Hits:Length() do
        local CurHit = Hits:Get(Ind)
        -- not check valid blocking hit as calc range may init penetrating with target.
        if CurHit.Component and not Components:Contains(CurHit.Component) then
            OutHits:Add(CurHit)
            Components:Add(CurHit.Component)
        end
    end

    return OutHits
end

-- DEPRECATED! Filter moved to CalcComponent
function HiCollisionLibrary.CheckFilterAndLimit(Hits, TargetFilter, CalcTargetLimit)
    local OutHits = UE.TArray(UE.FHitResult)
    local HitActors = UE.TArray(UE.AActor)
    for Ind = 1, Hits:Length() do
        local CurHit = Hits:Get(Ind)
        local CurComp = CurHit.Component
        local CurActor = CurComp:GetOwner()
        if CalcTargetLimit > 0 and HitActors:Length() >= CalcTargetLimit then
            break
        end

        -- TODO IsDead check should be in TargetFilter.
        if CurActor and TargetFilter:FilterActor(CurActor)
                and HitActors:Find(CurActor) == 0 then
            OutHits:AddUnique(CurHit)
            HitActors:AddUnique(CurActor)
        end
    end

    return OutHits
end

return HiCollisionLibrary
